core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt

```diff
@@ -457,37 +459,76 @@ object UnitActions {
+
+    private fun addSpreadReligionActions(unit: MapUnit, actionList: ArrayList<UnitAction>, city: CityInfo, maxSpreadUses: Int) {
+        val blockedByInquisitor =
+            city.getCenterTile()
+                .getTilesInDistance(1)
+                .flatMap { it.getUnits() }
+                .any {
+                    it.hasUnique("Prevents spreading of religion to the city it is next to")
+                    && it.religion != unit.religion
+                }
         actionList += UnitAction(UnitActionType.SpreadReligion,
             title = "Spread [${unit.religion!!}]",
             action = {
-                unit.abilityUsedCount["Religion Spread"] = unit.abilityUsedCount["Religion Spread"]!! + 1
+                unit.abilityUsedCount[Constants.spreadReligionAbilityCount] = unit.abilityUsedCount[Constants.spreadReligionAbilityCount]!! + 1
                 val followersOfOtherReligions = city.religion.getFollowersOfOtherReligionsThan(unit.religion!!)
                 for (unique in unit.civInfo.getMatchingUniques("When spreading religion to a city, gain [] times the amount of followers of other religions as []")) {
                     unit.civInfo.addStat(Stat.valueOf(unique.params[1]), followersOfOtherReligions * unique.params[0].toInt())
                 }
                 city.religion.addPressure(unit.religion!!, unit.getPressureAddedFromSpread())
                 unit.currentMovement = 0f
-                if (unit.abilityUsedCount["Religion Spread"] == maxReligionSpreads) {
-                    addGoldPerGreatPersonUsage(unit.civInfo)
+                if (unit.abilityUsedCount[Constants.spreadReligionAbilityCount] == maxSpreadUses) {
+                    if (unit.isGreatPerson())
+                        addGoldPerGreatPersonUsage(unit.civInfo)
                     unit.destroy()
                 }
-            }.takeIf { unit.currentMovement > 0 } 
+            }.takeIf { unit.currentMovement > 0 && !blockedByInquisitor } 
+        )
+    }
+    
+    private fun addRemoveHeresyActions(unit: MapUnit, actionList: ArrayList<UnitAction>, city: CityInfo, maxHerseyUses: Int) {
+        if (city.civInfo != unit.civInfo) return
+        // Only allow the action if the city actually has any foreign religion
+        if (city.religion.getPressures().none { it.key != unit.religion!! }) return
+        actionList += UnitAction(UnitActionType.RemoveHeresy,
+            title = "Remove Heresy",
+            action = {
+                unit.abilityUsedCount[Constants.removeHeresyAbilityCount] = unit.abilityUsedCount[Constants.removeHeresyAbilityCount]!! + 1
+                city.religion.removeAllPressuresExceptFor(unit.religion!!)
+                unit.currentMovement = 0f
+                if (unit.abilityUsedCount[Constants.removeHeresyAbilityCount] == maxHerseyUses) {
+                    if (unit.isGreatPerson())
+                        addGoldPerGreatPersonUsage(unit.civInfo)
+                    unit.destroy()
+                }
```

Comment:
This part (plus the "add 1 to uses") would be the same for all actions with X uses. Maybe should be a function, addUse(unit, ability).